src/source/src/manager.rs

```diff
@@ -252,23 +277,25 @@ impl SourceManager for MemSourceManager {
         let source_columns = columns.iter().map(SourceColumnDesc::from).collect();
         let source = SourceImpl::TableV2(TableSourceV2::new(columns));
 
-        // Table sources do not need columns and format
-        let desc = SourceDesc {
-            source: Arc::new(source),
-            columns: source_columns,
-            format: SourceFormat::Invalid,
-            row_id_index: Some(0), // always use the first column as row_id
-        };
+        let gen_source_desc: SourceDescAsyncClosure = Box::new(move || Box::pin(async {
+            // Table sources do not need columns and format
+            Ok(SourceDesc {
+                source: Arc::new(source),
+                columns: source_columns,
+                format: SourceFormat::Invalid,
+                row_id_index: Some(0), // always use the first column as row_id
+            })
+        }));
 
-        sources.insert(*table_id, desc);
+        sources.insert(*table_id, gen_source_desc);
         Ok(())
     }
 
-    fn get_source(&self, table_id: &TableId) -> Result<SourceDesc> {
+    fn get_source(&self, table_id: &TableId) -> Result<&SourceDescAsyncClosure> {
```

Comment:
It seems multiple source instances will be created if we call `get_source` and `await` on the future multiple times. Is this expected? Maybe you need something like `tokio::sync::OnceCell`.

Reply:
it is possible that one source is used in multiple dags so I hope `&SourceDescAsyncClosure` can act as a factory func and for one closure, it can produce more than one identical source instances.